kernel/src/hil/hmac.rs

```diff
@@ -0,0 +1,39 @@
+//! Interface for HMAC
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+pub trait HMAC<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Set the hash key
+    fn set_key(&self, key: *const u8, len: usize) -> ReturnCode;
+
+    /// Add data to the hmac IP.
+    /// All data passed in is fed to the HMAC hardware block.
+    /// Returns the number of bytes written on success
+    fn add_data(&self, data: *const u8, length: usize) -> Result<usize, ReturnCode>;
```

Comment:
This may be an asynchronous operation on many hardware platforms and therefore should return a callback to the client. For efficiency reasons it may be desirable to not keep the scheduler busy with callbacks, so you could use the return value to indicate whether a client must wait for a callback.

Reply:
It is asyncronous here (not in the actual OT implementation due to a bug though). We call-back internally in the driver, kind of similar to the UART implementation.